File: backend/src/main/java/org/pupille/backend/mysql/screening/TerminDTOWithFilmDTOOverviewSemester.java
package org.pupille.backend.mysql.screening;

import org.pupille.backend.mysql.film.Film;
import org.pupille.backend.mysql.reihe.Reihe;
import org.pupille.backend.mysql.reihe.ReiheDTOGallery;
import org.pupille.backend.mysql.termin.Termin;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public record TerminDTOWithFilmDTOOverviewSemester(
        Long tnr,
        LocalDateTime vorstellungsbeginn,
        String titel,
        String terminBesonderheit,
        List<FilmDTOOverviewSemester> mainfilms,
        Set<ReiheDTOGallery> reihen,
        Integer terminGesamtlaufzeit
) {
    public TerminDTOWithFilmDTOOverviewSemester(Termin termin, List<Film> films, Set<Reihe> reihen, Integer terminGesamtlaufzeit) {
        this(
                termin.getTnr(),
                termin.getVorstellungsbeginn(),
                termin.getTitel(),
                termin.getBesonderheit(),

                films.stream()
                        .map(FilmDTOOverviewSemester::new)
                        .toList(),

                reihen.stream()
                        .map(ReiheDTOGallery::new)
                        .collect(Collectors.toSet()),

                terminGesamtlaufzeit
        );
    }
}

File: frontend/src/types/TerminDTOWithFilmDTOOverviewSemester.ts
import FilmDTOOverviewSemester from "./FilmDTOOverviewSemester.ts";
import ReiheDTOGallery from "./ReiheDTOGallery.ts";

// this one mirrors TerminDTOWithFilmDTOOverviewSemester
// for entries on the OverviewSemester component
interface TerminDTOWithFilmDTOOverviewSemester{
    tnr: number;
    vorstellungsbeginn?: string;
    titel?: string | null;
    terminBesonderheit?: string | null;
    mainfilms: FilmDTOOverviewSemester[];
    reihen: ReiheDTOGallery[];
    terminGesamtlaufzeit: number;
}

export default TerminDTOWithFilmDTOOverviewSemester;